### DISCUSSION TITLE
Nginix and MQTT TLS

### DISCUSSION THREAD

**Discussion by DonatoD** [arendst/Tasmota#15233] [Q&A] [answered]
Hi, It's more than one week I'm struggling with mqtt TLS communication.
Using gitpod I built my own .bin file and was able to flash my shelly 2.5. Everythings works great, but my mqtt connection which requires TLS.
My mosquitto is behind an nginx reverse proxy and it works great with node-red and others software. Just flag TLS box and there is it. My certificate is Let's Encrypt.
On Tasmota console I always get
```
Failed to connect to www.myexample.com:2883, rc -2. Retry in 10 sec
QPC: Reset
MQT: Attempt to connect...
MQT: TLS connection error: 296
```
while, looking at nginix log I get:
```
[info] 29113#29113: *197 client 5.92.36.212:54054 connected to 0.0.0.0:2883
 [info] 29113#29113: *197 SSL_do_handshake() failed (SSL: error:1417A0C1:SSL routines:tls_post_process_client_hello:no shared cipher) while SSL ha>
```
this is my last nginix configuration under /etc/nginx/stream_conf.d/
```
log_format mqtt '$remote_addr [$time_local] $protocol $status $bytes_received '
                '$bytes_sent $upstream_addr';
upstream msqt_mqtt {
    server 127.0.0.1:1883;
}
server {
    listen 2883 ssl;
    proxy_pass msqt_mqtt;
    ssl_certificate /etc/nginx/ssl/myexample.com/fullchain.cer;
    ssl_certificate_key /etc/nginx/ssl/myexample.com/myexample.key;
    ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
    ssl_prefer_server_ciphers on;
    ssl_ciphers ECDHE-RSA-AES128-GCM-SHA2566,ECDH+AESGCM:DH+AESGCM:ECDH+AES256:DH+AES256:ECDH+AES128:DH+AES:ECDH+3DES:DH+3DES:RSA+AESGCM:RSA+AES:RSA+3DES:!aNULL:!MD5>
    ssl_ecdh_curve secp384r1;
    ssl_session_tickets off;
    access_log /var/log/nginx/mqtt_access.log mqtt;
    error_log  /var/log/nginx/mqtt_error.log info;
}
```

Thanks a lot

**Comment by joba-1** [comment]
the error message seems pretty clear: no common ciphers. Have you checked this? 

I don't know which ciphers tasmota supports, but it is only a selected few due to resource constraints.

And while you are at it: go to a ssl security site and check for recommended ciphers and tls versions. Pretty sure it is only tls1.2 and tls1.3 while your config lists older ones.

**Reply by sfromis** [reply]
Did you read the [Tasmota TLS docs](https://tasmota.github.io/docs/TLS/) with information about what's supported? This includes "RSA 2048 bits SHA 256", simplest is to use fingerprint validation.

**Reply by DonatoD** [reply]
Thanks, yes I read the Tasmota TLS docs and from the message I focused on the certificate.  But I still cannot undersand how to solve the problem. From the document I modified this line adding the  ssl_ciphers required (as you can see in nginix configuration)
`ssl_ciphers ECDHE-RSA-AES128-GCM-SHA2566, ....`
I even deleted TLSv1 and TLSv1.1 as suggested by  @joba-1. Now the line is
`ssl_protocols TLSv1.2 TLSv1.3;`
But nothing. Regarfind fingerprint I understand its not needed with let's encrypt certificate.
Forgive me but this is my first time with Tasmota.
Thanks for the support

**Reply by sfromis** [reply]
Based on my understanding (without having used it), my starting point would be to always go for fingerprint validation, unless it is somehow needed to go to the trouble of CA validation.

**Reply by DonatoD** [reply]
So, in order to try with fingerprint I should flesh the device again, correct?
Using gitpod how should I place the line
`#define MQTT_FINGERPRINT1     "00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00"`
For enabling mqtt TLS I could just choose it

**Reply by s-hadinger** [reply]
Just to clarify, your server must support TLS 1.2 and ECDHE-RSA-AES128-GCM-SHA256

**Reply by DonatoD** [reply]
thanks @s-hadinger. It's not enough the declaration in my nginx configuration? Maybe there is something wrong with it. I posted it at first message

**Reply by s-hadinger** [reply]
I just realized there is a typo in your cipher. It should be ECDHE-RSA-AES128-GCM-SHA256 and not ECDHE-RSA-AES128-GCM-SHA2566

**Reply by bogorad** [reply]
> I just realized there is a typo in your cipher. It should be ECDHE-RSA-AES128-GCM-SHA256 and not ECDHE-RSA-AES128-GCM-SHA2566

it's from the official doc, and the bug is still there

https://tasmota.github.io/docs/TLS/#limitations

**Reply by barbudor** [reply]
@bogorad you are welcomed to submit a change to the docs by clicking on the pencil button on the top left
Thanks for your contribution

**Reply by bogorad** [reply]
> @bogorad you are welcomed to submit a change to the docs 

I did and it was accepted already, thank you!

The issue (in my case at least) was that Certbot (starting with v2.0.0) is requesting an EC-key, not an RSA-key by default.

**Comment by DonatoD** [comment, marked as answer]
No, it's still continue getting the same error. I made several test but nothing. I even deleted all the others ssl_ciphers settings in order to avoid any kind of typo, but the problem remain.
Could you please guide me to check if my certificate is compliant with Tasmota requirement. For what I've been able to verify, yes. But I'm not that expert with this stuff.
Thanks again for the support

**Reply by DonatoD** [reply]
@sfromis Using openssl command I get this result. Could you please help me to understand if the certificate is complient with Tasmota request?

```
issuer=C = AT, O = ZeroSSL, CN = ZeroSSL RSA Domain Secure Site CA

---
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: RSA-PSS
Server Temp Key: ECDH, P-384, 384 bits
---
SSL handshake has read 5555 bytes and written 753 bytes
Verification: OK
---
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Server public key is 2048 bit
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
---
Post-Handshake New Session Ticket arrived:
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : TLS_AES_256_GCM_SHA384
    Session-ID: 93B083AA3B3BFE3548C5322C8895B332D1F165C2ADA561EE4D2AB030789D19FE
    Session-ID-ctx:
    Resumption PSK: 2BCB38274754C01D13FC87E2FB9B31821CE553787BE05A4BD302179C8080C1E09A16F896A8860FCCF04E599428DC2877
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 300 (seconds)
    TLS session ticket:
    0000 - 6f 9d 86 91 5c a6 9f 3e-f1 da ed c6 9b e1 83 9a   o...\..>........
    0010 - b7 f8 dd fc 5d d0 fd c4-e9 79 e2 5e 57 59 c6 b9   ....]....y.^WY..

    Start Time: 1648494902
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
    Max Early Data: 0
---
read R BLOCK
---
Post-Handshake New Session Ticket arrived:
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : TLS_AES_256_GCM_SHA384
    Session-ID: 7ABAC0F05EB75CA8F971BF90394342F4F78096EA79D94E8CC291B2A78912D92C
    Session-ID-ctx:
    Resumption PSK: E15C9D6DB03CC85DF50E5067BEBE1D0BD6F1A7B2DF56EB2C4112962727A56E448FBD08C158876F8DDECF540F5DB1F645
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 300 (seconds)
    TLS session ticket:
    0000 - af 2b 6f 47 25 97 e8 8d-d6 e9 19 e0 32 cb c8 fe   .+oG%.......2...
    0010 - d3 a5 00 0f 00 a2 20 2a-bb 72 85 5a 97 38 d7 d3   ...... *.r.Z.8..

    Start Time: 1648494902
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
    Max Early Data: 0

```

**Reply by s-hadinger** [reply]
What openssl command did you use? This is not your server's certificate, but information about an openssl session, which is not using the same parameters as Tasmota, hence we cannot infer anything from it.

Also, the `no_shared_cipher` is independent from the server's certificate.

There must be options in openssl to force TLS1.2 and Tasmota's cipher if you want to try it.

**Reply by redhat421** [reply]
Might give this a shot to see if it fails:

`% openssl s_client -ciphersuites TLS_AES_128_GCM_SHA256 -connect server.name:8883`

Since you're connecting using a ZeroSSL cert, I suspect the problem is that you need SHA384 support in Tasmota which requires a patch.

Thanks!

**Reply by redhat421** [reply]
Here is an example change which enables SHA384 hashes:
https://github.com/redhat421/Tasmota/commit/bad89a8b49803e4198bd790ea04ed0e9f6683159

Thanks!

**Reply by DonatoD** [reply]
@s-hadinger the command I used was: openssl s_client -connect myexample.com:2883 -showcerts
Now, with the command @redhat421 suggested, this is the result:
openssl s_client -ciphersuites TLS_AES_128_GCM_SHA256 -connect myexample.com:2883

```
CONNECTED(00000003)
depth=2 C = US, ST = New Jersey, L = Jersey City, O = The USERTRUST Network, CN = USERTrust RSA Certification Authority
verify return:1
depth=1 C = AT, O = ZeroSSL, CN = ZeroSSL RSA Domain Secure Site CA
verify return:1
depth=0 CN = myexample.com
verify return:1
---
Certificate chain
 0 s:CN = myexample.com
   i:C = AT, O = ZeroSSL, CN = ZeroSSL RSA Domain Secure Site CA
 1 s:C = AT, O = ZeroSSL, CN = ZeroSSL RSA Domain Secure Site CA
   i:C = US, ST = New Jersey, L = Jersey City, O = The USERTRUST Network, CN = USERTrust RSA Certification Authority
 2 s:C = US, ST = New Jersey, L = Jersey City, O = The USERTRUST Network, CN = USERTrust RSA Certification Authority
   i:C = GB, ST = Greater Manchester, L = Salford, O = Comodo CA Limited, CN = AAA Certificate Services
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIGeTCCBGGgAwIBAgIQWQmFNFB4TJA7eOasLaqIBTANBgkqhkiG9w0BAQwFADBL
........................
........................
QQEiAWAFFnobTrOHAyu3uWtqFVLCG1vTHHQLyVc=
-----END CERTIFICATE-----
subject=CN = myexample.com

issuer=C = AT, O = ZeroSSL, CN = ZeroSSL RSA Domain Secure Site CA

---
No client certificate CA names sent
Peer signing digest: SHA256
Peer signature type: RSA-PSS
Server Temp Key: ECDH, P-384, 384 bits
---
SSL handshake has read 5539 bytes and written 729 bytes
Verification: OK
---
New, TLSv1.3, Cipher is TLS_AES_128_GCM_SHA256
Server public key is 2048 bit
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
---
Post-Handshake New Session Ticket arrived:
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : TLS_AES_128_GCM_SHA256
    Session-ID: FBD0D08296D832296CDF645D026D69877C8C2F3E503FD61D4E2DB18E5B47DCBC
    Session-ID-ctx:
    Resumption PSK: ECE8DBBD164D5A76A1B7C3085776C658F9FC1215C9A95F08995B58095A6D58A0
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 300 (seconds)
    TLS session ticket:
    0000 - 8b 65 42 d1 7b 19 04 93-c1 3e b0 2f d7 26 18 af   .eB.{....>./.&..
    0010 - 9a de 12 f7 7e b6 c3 82-00 a2 94 c1 e7 49 a2 a2   ....~........I..

    Start Time: 1648568356
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
    Max Early Data: 0
---
read R BLOCK
---
Post-Handshake New Session Ticket arrived:
SSL-Session:
    Protocol  : TLSv1.3
    Cipher    : TLS_AES_128_GCM_SHA256
    Session-ID: E236319932773C0177D20C3FAB06CFD043E7C28CDC898A5F78D308A65F4B52ED
    Session-ID-ctx:
    Resumption PSK: 903A080F863C77C6CD1EFEB5ECB8642967F97EB0473825D5986EA0177E7BB9E3
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    TLS session ticket lifetime hint: 300 (seconds)
    TLS session ticket:
    0000 - df 58 98 1e a7 6b 53 4d-19 8a a6 a8 16 5e 9a a5   .X...kSM.....^..
    0010 - da df 89 a8 29 82 16 89-b5 20 d4 9c 18 d4 97 27   ....).... .....'

    Start Time: 1648568356
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
    Max Early Data: 0
---
read R BLOCK

```

**Reply by s-hadinger** [reply]
Hmmm. I don't see any issue here. Tasmota uses P256 curve for ECDHE instead of P384 but it should be supported by your server, and it's not linked to the certificate (which is RSA 2048).

I'm just curious how openssl can connect using TLS 1.3, if you haven't defined it in your server.

**Reply by DonatoD** [reply]
I do not know what to say!!! What I've learned is that I'm not using let's encrypt as I thought, but ZeroSSL. I followed this link https://github.com/acmesh-official/acme.sh where in fact zeroSSl is the default. Maybe the problem is this one as @redhat421 suspects

**Reply by redhat421** [reply]
I'll bet you a shiny nickel the problem is that you don't have `SHA384` in your Tasmota build.

Add that to `WiFiClientSecureLightBearSSL.cpp` and I'll bet it will work.

Good luck!

**Reply by redhat421** [reply]
Alternately, just switch to using Letsencrypt certs. :-D

**Reply by DonatoD** [reply]
Sorry @s-hadinger. Maybe something has escaped within the meassages. As suggested, I made this change within ngnix configuration for mqtt:

```
ssl_ciphers ECDHE-RSA-AES128-GCM-SHA256;
ssl_protocols TLSv1.2 TLSv1.3;
```

**Reply by DonatoD** [reply]
@redhat421 yes I was just thinking the same and googling for some help on how to do it

**Reply by DonatoD** [reply]
I was even looking at the link you posted regarding an example on how to enable SHA384 hashes. But I don't know how to make this change. I 'm able to use tasmocompiler []https://github.com/benzino77/tasmocompiler for building tasmota, but from the site you suggest I don't know which change to make.

**Reply by redhat421** [reply]
Ha! If you didn't enable the use of ZeroSSL certs on purpose, then your CA validation won't be right either (The chain of trust is hardcoded for Amazon and Letsencrypt certs). If you're simply planning on doing fingerprint validation, you might as well use self signed certs.

Good luck!

**Reply by s-hadinger** [reply]
Would it be possible for your server to be accessible for tests over the internet?

**Reply by DonatoD** [reply]
Thanks so much @s-hadinger. We can arrange a session via anydesk, so we could even make some check on the Tasmota device and even modify it if needed.
In the meanwhile I'm doing other test without success. I even try with  digicert certificate, but nothing.
What I've noticed is that no matter what certificate I use, when I set
1) "ssl_ecdh_curve secp521r1" within nginx configuration I get 
- on Tasmota console
TLS connection error: 296
- On nginix log
2022/03/31 19:46:45 [info] 748#748: *434 client 5.92.47.12:62505 connected to 0.0.0.0:2883 
2022/03/31 19:45:34 [info] 748#748: *433 SSL_do_handshake() failed (SSL: error:1417A0C1:SSL routines:tls_post_process_client_hello:no shared cipher) while SSL handshaking, client: 5.92.47.12, server: 0.0.0.0:2883
2) otherwise
- on Tasmota console
TLS connection error: 62
- On nginix log
2022/03/31 19:53:10 [info] 11708#11708: *439 client 5.92.47.12:62546 connected to 0.0.0.0:2883
2022/03/31 19:53:10 [info] 11708#11708: *439 peer closed connection in SSL handshake (104: Connection reset by peer) while SSL handshaking, client: 5.92.47.12, server: 0.0.0.0:2883

**Reply by redhat421** [reply]
Looking at the Tasmota side errors for the second case:
https://www.bearssl.org/apidoc/bearssl__x509_8h.html

`BR_ERR_X509_NOT_TRUSTED   62`
`X.509 status: chain could not be linked to a trust anchor.`

Looks like the CA / Trust issues are biting you in the second case. I think the Letsencrypt cert baked in is for the `rsa` cypher, so make sure you're requesting that in your acme client.

Thanks!

**Reply by DonatoD** [reply]
@s-hadinger if you can't with anydesk, just let me know when you're available and I'll arrange to give you access for all a day long.
Again, thanks so much

@redhat421 thnaks for your support, too. I'm trying and googlig for help but nothing, I'm not able to get it :-(

**Reply by redhat421** [reply]
I still don't see the Letsencrypt cert in any of your examples.

You should request an RSA cert from Letsencrypt.
I'll bet if you dropped such a cert in, Tasmota would happily connect. :-)

For debugging, you can also do things like:
```
openssl s_server -msg -state -tls1_2 -tlsextdebug  -key privkey.pem -cert cert.pem -cert_chain fullchain.pem -port 8884
```

**Reply by s-hadinger** [reply]
Hold on. Why are you using `"ssl_ecdh_curve secp521r1"`?

This cipher is definitely not supported. Tasmota only supports P256.

You should use `"ssl_ecdh_curve prime256v1"` instead

**Reply by DonatoD** [reply]
Thanks @s-hadinger  If I put ssl_ecdh_curve prime256v1 I get this error:
- on Tasmota console
TLS connection error: 62
- On nginix log
2022/03/31 19:53:10 [info] 11708#11708: *439 client 5.92.47.12:62546 connected to 0.0.0.0:2883
2022/03/31 19:53:10 [info] 11708#11708: *439 peer closed connection in SSL handshake (104: Connection reset by peer) while SSL handshaking, client: 5.92.47.12, server: 0.0.0.0:2883

**Reply by DonatoD** [reply]
Thanks @redhat421 with Let's Encrypt everything works great. I reinstalled everything following the main site. As Tasmota configurtion I added
```
#ifndef USE_MQTT_TLS
  #define USE_MQTT_TLS
  #define USE_MQTT_TLS_CA_CERT
  #define USE_MQTT_TLS_FORCE_EC_CIPHER
  #define MQTT_PORT 2883
#endif
```

My server provider offers one year DigiCert certificate included in the server subscription, and for me it would be the top to be able to use it. I would be grateful if someone could help me to understand the characteristics of my DigiCert certificate and which Tasmota/Nginix configuration I need. With other programs, like node-red, it works like a charm with no particular settings.
Thanks so much to all

**Reply by redhat421** [reply]
Glad everything is working now. If you have auto cert renewal already managing your Letsencrypt certs, that's not a bad way to leave things in my opinion.

If you want to use another CA (such as DigiCert), you can follow this guide, only include the chain of trust for DigiCert instead of the self signed DIY CA that guide recommends. Here is the link to that instruction:
https://tasmota.github.io/docs/Self-signed-Mosquitto/#2-configure-your-tasmotabuild-on-compiling-machine

Note that having DigiCert sign your MQTT server cert doesn't really buy you that much in this case (since you're not trying to impress any web browsers). I'd consider running the above linked guide as is, and save yourself the trouble of dealing with any external CAs.

Good luck!

**Reply by DonatoD** [reply]
Thanks @redhat421. I don't understand how to include the chain of trust for DigiCert instead of the self signed DIY CA within user_config_override.h

**Reply by redhat421** [reply]
Looking at the [Self signed Mosquitto docs](https://tasmota.github.io/docs/Self-signed-Mosquitto/#2-configure-your-tasmotabuild-on-compiling-machine) I linked.

> Merge your myca.ino to $TASMOTAROOT/tasmota/tasmota_ca.ino. Result should look like:

Changes to CA trust will end up in the [`tasmota_ca.ino`](https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_ca.ino) file.

You'll need to find [DigiCert's CA certs](https://www.digicert.com/kb/digicert-root-certificates.htm) and run the command listed here: https://github.com/arendst/Tasmota/blob/089ace9d358149fec6a99e4f35998bd9bd13c063/tasmota/tasmota_ca.ino#L32

Once you have the DigiCert CA added to that file, it will be ready to validate any cert signed by that CA.

Since DigiCert has a bunch of CAs, the easiest way to do this would be to make the cert (ie ask DigiCert to sign your MQTT server's cert) and then inspect it to see which cert chain signed it. Once you know that, you'll know which cert to add.

Having said all that, I'd again encourage you to consider using a self signed cert, or just stick with the LetsEncrypt approach,  since I suspect both would be easier. 

Here's a list of pros and cons I could think up for each approach:

Pros for Letsencrypt:
  * Produces certs that are automatically trusted by other tools (eg in CA trust files).
  * Rotation is automatic (via certbot or the like).
  * Well known configuration, more help available if things go wrong.

Cons for Letsencrypt:
  * Monitoring required: If automation fails, your certs will run out of validity quickly (within 90 days).

Pros for Self signed:
  * Simpler (in some ways)
  * Very long validity times available. You can set the cert to expire in 2038.
  * No external dependencies. If Letsencrypt goes out of business (seems very unlikely) you're stuck having to update firmware on devices.
  * Much easier to do peer cert validation.    

Cons for Self signed:
  * No automatic trust. You need to add your signing CA to all of the hosts which you want to be trusted.
  * Possibly less secure, if you lose a cert which is set to expire in the far future, you'll may need to invalidate the CA and start over.

Pros for DigiCert:
  * Produces certs that are automatically trusted by other tools (eg in CA trust files).
  * Longer expiry times on generated certs (1 year vs 90 days for LetsEncrypt).

Cons for DigiCert:
  * No automatic rotation. You'll need to rotate the server cert every year.
  * Slightly more complex setup.
  * More unique setup, debugging could be a little bit harder.

Good luck!

**Reply by DonatoD** [reply]
Thanks for all support. Yes I agree with you that it's easier just keep let's Encrypt. But since I'm new with Tasmota and I bumped into this issue, I'm trying to understand better how it works, for sure it will help me for more other things.
Thanks so much

**Comment by DonatoD** [comment]
Excuse me, one more info about fingerprint, that I've not tested yet. The devices will be installed in places not so easy to reach, and will perform some action remotly, For this reason I need to limite as much as possible any chance of sending there people to flush it, update it, or whotever. What it's not completly clear to me is if in case of changing certificate authority, for example, will require to manually re-flush it in order to put firngerprint in auto-learning again or not (from Tasmota site I read that renewing certificate won't require any action).

**Reply by sfromis** [reply]
In case of needing to execute commands, that is possible while you have a MQTT connection, and you could set up rules to have commands executed after boot, when they can deactivate the rule after having done that. Of source, any such setup change needs to be done with great care to mitigate the risk of it ending up in a need to do local recovery.
